### DISCUSSION TITLE
Proposal for an Azure Cognitive Search Plugin

### DISCUSSION THREAD

**Discussion by zentooo** [danny-avila/LibreChat#567] [Feature Requests & Suggestions]
Hello LibreChat community,

I hope this message finds you well. I am writing to propose a new plugin for LibreChat, specifically designed to integrate with Azure Cognitive Search. I believe this addition would greatly enhance our ability to retrieve and interact with closed information within an organization using LibreChat.

The primary purpose of this plugin would be to facilitate keyword-based searches of a company's internal data using Azure Cognitive Search. This is somewhat different from the Vector Search functionality discussed in a previous discussion (https://github.com/danny-avila/LibreChat/discussions/431). While Vector Search undoubtedly has its benefits, the focus of this proposed plugin is on simple keyword search capability, which better suits my needs.

In terms of implementation, I envision this Azure Cognitive Search Plugin to work similarly to the existing Google search plugin that we currently have in LibreChat. The aim is to provide a straightforward and effective way to search internal documents and data.

Should this proposal be accepted, I am ready and willing to write the code and submit a Pull Request. I believe that this Azure Cognitive Search Plugin would significantly improve the versatility and usefulness of LibreChat, especially for users who rely on Azure for their internal data storage and retrieval needs.

Thank you for considering this proposal. I look forward to hearing your thoughts and feedback.

**Comment by fuegovic** [comment]
Hi @zentooo , this seems like a very interesting feature, I personally see no issue with your proposal

**Reply by zentooo** [reply]
Hi @fuegovic, thanks for your input! I appreciate your support for the proposal.

**Comment by danny-avila** [comment]
Hi @zentooo, it should be fairly simple if there is already an implementation of azure cognitive search with javascript, in any capacity. I haven't looked yet but I'm sure there is.

I would like to add that it would be possible for me to develop if I had azure openAI and cognitive search setup but I'm lacking in either case.

**Reply by zentooo** [reply]
Hi @danny-avila, indeed, there's an official JavaScript library for Azure Cognitive Search that could simplify things quite a bit ([official documentation](https://learn.microsoft.com/en-us/javascript/api/overview/azure/search-documents-readme?view=azure-node-latest)).

Also, I have access to an Azure environment. But I'm wondering, would it be better if I refrained from taking this on?

**Reply by danny-avila** [reply]
if you would like to take this on, I highly encourage it! 

You can read my docs on making a plugin compatible for this project here: https://docs.librechat.ai/features/plugins/make_your_own.html

I would then add that you check the new structured tool that is compatible with openAI functions, which will be the de facto manner of creating these langchain tools. api/app/langchain/tools/structured/StableDiffusion.js (location may change soon as I'm refactoring and writing up the new documentation)

**Reply by zentooo** [reply]
Thanks for your encouragement and the resources! I'll start delving into the development with reference to your documentation.

As for the new structured tool, if it's compatible only with APIs supporting function calls, do you think we should also prepare non-structured tools for APIs that aren't compatible?
